staging/src/k8s.io/component-base/logs/json/json.go

```diff
@@ -0,0 +1,190 @@
+		// fast-return if we have no suggared fields.
+		return additional
+	}
+
+	// unlike Zap, we can be pretty sure users aren't passing structured
+	// fields (since logr has no concept of that), so guess that we need a
+	// little less space.
+	fields := make([]zap.Field, 0, len(args)/2+len(additional))
+	for i := 0; i < len(args); {
+		// check just in case for strongly-typed Zap fields, which is illegal (since
+		// it breaks implementation agnosticism), so we can give a better error message.
+		if _, ok := args[i].(zap.Field); ok {
+			l.DPanic("strongly-typed Zap Field passed to logr", zap.Any("zap field", args[i]))
+			break
+		}
+
+		// make sure this isn't a mismatched key
+		if i == len(args)-1 {
+			l.DPanic("odd number of arguments passed as key-value pairs for logging", zap.Any("ignored key", args[i]))
+			break
+		}
+
+		// process a key-value pair,
+		// ensuring that the key is a string
+		key, val := args[i], args[i+1]
+		keyStr, isString := key.(string)
+		if !isString {
+			// if the key isn't a string, DPanic and stop logging
+			l.DPanic("non-string key argument passed to logging, ignoring all later arguments", zap.Any("invalid key", key))
+			break
+		}
+
+		fields = append(fields, zap.Any(keyStr, val))
+		i += 2
+	}
+
+	return append(fields, additional...)
+}
+
+// zapLogger is a logr.Logger that uses Zap to log.
+type zapLogger struct {
+	// NB: this looks very similar to zap.SugaredLogger, but
+	// deals with our desire to have multiple verbosity levels.
+	l *zap.Logger
+	infoLogger
+}
+
+var _ logr.Logger = &zapLogger{}
+
+// Error write log message to error level
+func (l *zapLogger) Error(err error, msg string, keysAndVals ...interface{}) {
+	if checkedEntry := l.l.Check(zap.ErrorLevel, msg); checkedEntry != nil {
+		checkedEntry.Time = timeNow()
+		checkedEntry.Write(handleFields(l.l, keysAndVals, handleError(err))...)
+	}
+}
+
+// V return info logr.Logger with specified level
+func (l *zapLogger) V(level int) logr.InfoLogger {
+	lvl := zapcore.Level(-1 * level)
+	if l.l.Core().Enabled(lvl) {
+		l.lvl = lvl
+		return l
+	}
+	return disabledInfoLogger
+}
+
+// WithValues return logr.Logger with some keys And Values
+func (l *zapLogger) WithValues(keysAndValues ...interface{}) logr.Logger {
+	newLogger := l.l.With(handleFields(l.l, keysAndValues)...)
+	return NewJsonLogger(newLogger, nil)
+}
+
+// WithName return logger Named with specified name
+func (l *zapLogger) WithName(name string) logr.Logger {
+	newLogger := l.l.Named(name)
+	return NewJsonLogger(newLogger, nil)
+}
+
+// NewJsonLogger creates a new json logr.Logger using the given Zap Logger to log.
+func NewJsonLogger(l *zap.Logger, w zapcore.WriteSyncer) logr.Logger {
+	encoderConfig := zapcore.EncoderConfig{
+		MessageKey: "msg",
+
+		LevelKey:    "v",
+		EncodeLevel: int8LevelEncoder,
+
+		TimeKey:    "ts",
+		EncodeTime: zapcore.EpochMillisTimeEncoder,
+	}
+
+	var syncer zapcore.WriteSyncer
+	syncer = os.Stdout
+	if w != nil {
+		syncer = w
+	}
```

Comment:
```suggestion
	if w != nil {
		w = os.Stdout
	}
```

Reply:
If w is set, will add a syncer, with w, to logger, that's allow set another output for jsonlogger.

Reply:
Yes, so basically `w` default value should be `os.Stdout`
I propose to use `w` instead of creating additional variable `syncer`. That way we reduce number of variables and make it more obvious that this code implements "default pattern".

Reply:
I got this.